### DISCUSSION TITLE
cache_miss on every call to pjitted function

### DISCUSSION THREAD

**Discussion by hr0nix** [jax-ml/jax#15915] [Q&A]
I've been looking at traces of model-parallel training I've implemented in jax using pjit and noticed a curious thing: every call to pjitted function hits a function called `cache_miss` and does quite a lot of computations in python.

<img width="1625" alt="Screen Shot 2023-05-08 at 22 30 08" src="https://user-images.githubusercontent.com/4273011/236939495-d4053bf7-fece-498f-8248-75d2ba06e5f4.png">

I'm wondering whether this is expected or I have set up something wrong and jax is re-doing some work it's supposed to do only once. It has practical significance for me because overhead from these python pjit activities can sometimes be large enough to make GPU pipeline wait.

The training step function I pjit looks something like this:
```
def training_step(
        train_state: TrainState,
        batch: PyTree,
) -> tuple[TrainState, jax.Array]:
    rng, new_rng = jax.random.split(train_state.rng)

    loss_value_and_grad_fn = jax.value_and_grad(compute_loss, argnums=0)
    loss, grad = loss_value_and_grad_fn(train_state.params, train_state, batch)

    updates, new_optimizer_state = train_state.tx.update(grad, train_state.opt_state, train_state.params)
    new_params = optax.apply_updates(train_state.params, updates)

    train_state = train_state.replace(
        params=new_params,
        opt_state=new_optimizer_state,
        rng=new_rng,
    )

    return train_state, loss
```

Here's how I compile it:
```
train_on_batch_fn = jax.experimental.pjit.pjit(
    training_step,
    in_shardings=(state_mesh_annotations, data_sharding),
    out_shardings=(state_mesh_annotations, None),
    donate_argnums=(0,),  # Donate train_state
)
```

On every training step I just call it like this:
```
train_state, loss = train_on_batch_fn(train_state, batch)
```

**Comment by quattro** [comment]
Any updates on this? I'm running into a similar issue.

**Reply by jakevdp** [reply]
Could you open a new issue with a reproduction? In general JIT recompilation will happen when the JIT cache key doesn't match previous calls. It will be much easier to narrow-down why it's happening if you can share your code.

**Reply by quattro** [reply]
Sure. I'll try my best--it is happening in a project's codebase, and I haven't been able to reproduce it with a MWE at the moment. Hopefully I'll have time either later this week or next week to circle back to it.

**Reply by jakevdp** [reply]
If a MWE is difficult, then you can set the `jax_log_compiles` flag to `True`, and then each time a function is compiled it will log relevant information about why it's being recompiled.

**Reply by quattro** [reply]
Thanks @jakevdp . This is what caused me to notice the issue. I'm having a very hard time replicating the issue in a MWE, but the issue is,
```python
  # self.proj is a BCOO
  proj_normalizer = self.proj @ normalizer # cache miss

  proj_counts = self.proj @ jnp.ones_like(normalizer) # cache miss
  is_zero = jnp.isclose(proj_normalizer, 0.0)
  where_zero = self.proj.T @ is_zero.astype(normalizer.dtype) # cache miss

  normalizer = jnp.where(where_zero != 0.0, 1.0, normalizer)
  proj_normalizer = jnp.where(is_zero, proj_counts, proj_normalizer)

  res = self.proj @ (vec * normalizer) # cache miss
  res = res / proj_normalizer
```
Each inner product with the BCOO (i.e. `self.proj`) is a cache miss, despite all of the right-hand-side variables being of same shape and dtype (i.e. `p` and `float64`).

For example, the cache miss info at the first miss is:

```
TRACING CACHE MISS at /Users/nicholas/Projects/sel-zero/src/sel_zero/dynamics/projection.py:54:26 (Projection.weighted_projection) because:
  never seen function:
    functools.partial(<function dot_general at 0x12d517240>, dimension_numbers=(((1,), (0,)), ((), ())), preferred_element_type=dtype('float64')) id=11219608224 defined at /Users/nicholas/micromamba/envs/jax/lib/python3.11/site-packages/jax/_src/lax/lax.py:739
  but seen another function defined on the same line; maybe the function is
  being re-defined repeatedly, preventing caching?
Finished tracing + transforming <unnamed wrapped function> for pjit in 0.00014710426330566406 sec
```

The repeated logged information at other misses is exactly the same, but with different ids for the partials. I've also tried pulling out and compiling separately a small function that is just the dot product between the BCOO and something else, but with no luck.

**Reply by jakevdp** [reply]
If the size of the underlying buffers in `self.proj` changes, then it will trigger a recompilation. Perhaps that's the issue here?

**Reply by quattro** [reply]
I can see that being the case for the first instance (or maybe also the transposed instance), but subsequent calls should still hit the lookup, rather than a miss. I'm wondering if it has something to do with the partial's that wrap `dot_general` having different ids, yet functionally being exactly the same. I'm not sure how to work around that.

**Reply by quattro** [reply]
Is the abstract eval the culprit? It seems like creating [these partials ](https://github.com/google/jax/blob/be59f6ec4748a3e55b832ee4b3e5715608a6d99a/jax/experimental/sparse/bcoo.py#L741)on the fly results in distinct functions.

**Reply by jakevdp** [reply]
Are any of your arrays sharded? Different shardings would lead to recompilation

**Reply by quattro** [reply]
No, they're all on the same CPU device. I'm beginning to think it may be the partial that wraps `lax.dot_general` for `eval_shape` inside the [bcoo_dot_general_abstract_eval at line 714](https://github.com/google/jax/blob/be59f6ec4748a3e55b832ee4b3e5715608a6d99a/jax/experimental/sparse/bcoo.py#L741).

That corresponds with the cache log:

```
TRACING CACHE MISS at /Users/nicholas/Projects/sel-zero/src/sel_zero/dynamics/projection.py:54:26 (Projection.weighted_projection) because:
  never seen function:
    functools.partial(<function dot_general at 0x12b887240>, dimension_numbers=(((1,), (0,)), ((), ())), preferred_element_type=dtype('float64')) id=10837044384 defined at /Users/nicholas/micromamba/envs/jax/lib/python3.11/site-packages/jax/_src/lax/lax.py:739
  but seen another function defined on the same line; maybe the function is
  being re-defined repeatedly, preventing caching?
Finished tracing + transforming <unnamed wrapped function> for pjit in 0.0001659393310546875 sec
TRACING CACHE MISS at /Users/nicholas/Projects/sel-zero/src/sel_zero/dynamics/projection.py:56:22 (Projection.weighted_projection) because:
  never seen function:
    functools.partial(<function dot_general at 0x12b887240>, dimension_numbers=(((1,), (0,)), ((), ())), preferred_element_type=dtype('float64')) id=6094598160 defined at /Users/nicholas/micromamba/envs/jax/lib/python3.11/site-packages/jax/_src/lax/lax.py:739
  but seen another function defined on the same line; maybe the function is
  being re-defined repeatedly, preventing caching?
Finished tracing + transforming <unnamed wrapped function> for pjit in 0.000164031982421875 sec
TRACING CACHE MISS at /Users/nicholas/Projects/sel-zero/src/sel_zero/dynamics/projection.py:58:21 (Projection.weighted_projection) because:
  never seen function:
    functools.partial(<function dot_general at 0x12b887240>, dimension_numbers=(((1,), (0,)), ((), ())), preferred_element_type=dtype('float64')) id=6094588160 defined at /Users/nicholas/micromamba/envs/jax/lib/python3.11/site-packages/jax/_src/lax/lax.py:739
  but seen another function defined on the same line; maybe the function is
  being re-defined repeatedly, preventing caching?
Finished tracing + transforming <unnamed wrapped function> for pjit in 0.0001881122589111328 sec
TRACING CACHE MISS at /Users/nicholas/Projects/sel-zero/src/sel_zero/dynamics/projection.py:65:14 (Projection.weighted_projection) because:
  never seen function:
    functools.partial(<function dot_general at 0x12b887240>, dimension_numbers=(((1,), (0,)), ((), ())), preferred_element_type=dtype('float64')) id=10834850848 defined at /Users/nicholas/micromamba/envs/jax/lib/python3.11/site-packages/jax/_src/lax/lax.py:739
  but seen another function defined on the same line; maybe the function is
  being re-defined repeatedly, preventing caching?
Finished tracing + transforming <unnamed wrapped function> for pjit in 0.00015616416931152344 sec
```

I'm not sure how to get around that, since the partials are needed to keep the static args for shape eval.

**Reply by jakevdp** [reply]
Yeah that makes sense – I think we could fix this by switching from using `jax.eval_shape(f, *args)` to using the new(ish) `jax.jit(f).eval_shape(*args)` because this API allows for static argnums & argnames, and that would let us avoid the partial evaulations that break the cache

Do you want to give that a try locally and see if it improves things?

**Reply by quattro** [reply]
Ah too cool. Replacing the offending line with 
```python
  out_aval = jax.jit(lax.dot_general, static_argnames=("dimension_numbers", "preferred_element_type")).eval_shape(
          jax.ShapeDtypeStruct(lhs_spinfo.shape, lhs_data.dtype),
          jax.ShapeDtypeStruct(rhs.shape, rhs.dtype),
          dimension_numbers=dimension_numbers,
          preferred_element_type=preferred_element_type)
``` 
fixed it. Thanks so much!

**Reply by jakevdp** [reply]
Awesome - would you like to put together a PR with that fix?
